packages/test/src/test-integration-workflows.ts

```diff
@@ -204,3 +209,52 @@ test('Worker allows heartbeating activities after shutdown has been requested',
   await worker.run();
   t.is(cancelReason, 'CANCELLED');
 });
+
+export async function conditionTimeout0Simple(): Promise<boolean | undefined> {
+  let validationTimerFired = false;
+  // Don't lower this value, otherwise, in CI, there is risk that the timer fire
+  // while the WFT is being processed, which would result in an UnhandledCommand.
+  workflow
+    .sleep(10_000)
+    .then(() => (validationTimerFired = true))
+    .catch((e) => console.log(e));
+
+  return await workflow.condition(() => validationTimerFired, 0);
+}
+
+test('Internal patches does not cause non-determinism error on replay', async (t) => {
+  const { createWorker, executeWorkflow } = helpers(t);
+  const worker = await createWorker({
+    // Disable workflow caching, to force replay after the condition's sleep
+    maxCachedWorkflows: 0,
+  });
+  await worker.runUntil(async () => {
+    try {
+      await executeWorkflow(conditionTimeout0Simple);
+      t.pass();
+    } catch (e) {
+      t.fail((e as Error).message);
+    }
+  });
+});
+
+test('Condition with timeout 0 does not block indefinitely', async (t) => {
+  const { createWorker, startWorkflow } = helpers(t);
+  const worker = await createWorker();
+  await worker.runUntil(async () => {
+    const handle = await startWorkflow(conditionTimeout0Simple);
+    t.false(await handle.result());
+    const history = await handle.fetchHistory();
+
+    const timerStartedEvents = history.events?.filter(({ eventType }) => eventType === EVENT_TYPE_TIMER_STARTED);
+    t.is(timerStartedEvents?.length, 2);
+    t.is(timerStartedEvents?.[0].timerStartedEventAttributes?.timerId, '1');
+    t.is(tsToMs(timerStartedEvents?.[0].timerStartedEventAttributes?.startToFireTimeout), 10000);
+    t.is(timerStartedEvents?.[1].timerStartedEventAttributes?.timerId, '2');
+    t.is(tsToMs(timerStartedEvents?.[1].timerStartedEventAttributes?.startToFireTimeout), 1);
+
+    const markersEvents = history.events?.filter(({ eventType }) => eventType === EVENT_TYPE_MARKER_RECORDED);
+    t.is(markersEvents?.length, 1);
+    t.is(markersEvents?.[0].markerRecordedEventAttributes?.markerName, CHANGE_MARKER_NAME);
+  });
+});
```

Comment:
TBH I don't think we even need this test, the fact that the workflow returns false should be enough to cover the functionality.